synapse/handlers/identity.py

```diff
@@ -85,43 +142,55 @@ def threepid_from_creds(self, creds):
             return None
 
         try:
-            data = yield self.http_client.get_json(
-                "https://%s%s"
-                % (id_server, "/_matrix/identity/api/v1/3pid/getValidated3pid"),
-                {"sid": creds["sid"], "client_secret": client_secret},
-            )
+            query_params = {"sid": creds["sid"], "client_secret": client_secret}
+            if use_v2:
+                query_params["id_access_token"] = id_access_token
+
+            data = yield self.http_client.get_json(url, query_params)
         except HttpResponseException as e:
+            if e.code == 404 and use_v2:
+                # This identity server is too old to understand Identity Service API v2
+                # Attempt v1 endpoint
+                logger.warn("Got 404 when POSTing JSON %s, falling back to v1 URL", url)
+                return (yield self.threepid_from_creds(creds, use_v2=False))
+
             logger.info("getValidated3pid failed with Matrix error: %r", e)
             raise e.to_synapse_error()
 
-        if "medium" in data:
-            return data
-        return None
+        return data if "medium" in data else None
```

Comment:
Can you move this into the try/except and then we can move the `return (yield self.threepid_from_creds(creds, use_v2=False))` out side the except clause